ui/src/dashboards/components/TableOptions.tsx

```diff
@@ -0,0 +1,205 @@
+    const {dataLabels} = this.props
+
+    return _.isEmpty(dataLabels)
+      ? [this.timeField]
+      : dataLabels.map(label => {
+          const existing = this.fieldNames.find(f => f.internalName === label)
+          return (
+            existing || {internalName: label, displayName: '', visible: true}
+          )
+        })
+  }
+
+  public handleChooseSortBy = (option: Option) => {
+    const {tableOptions, handleUpdateTableOptions} = this.props
+    const sortBy = {
+      displayName: option.text === option.key ? '' : option.text,
+      internalName: option.key,
+      visible: true,
+    }
+
+    handleUpdateTableOptions({...tableOptions, sortBy})
+  }
+
+  public handleTimeFormatChange = timeFormat => {
+    const {tableOptions, handleUpdateTableOptions} = this.props
+    handleUpdateTableOptions({...tableOptions, timeFormat})
+  }
+
+  public handleToggleVerticalTimeAxis = verticalTimeAxis => () => {
+    const {tableOptions, handleUpdateTableOptions} = this.props
+    handleUpdateTableOptions({...tableOptions, verticalTimeAxis})
+  }
+
+  public handleToggleFixFirstColumn = () => {
+    const {handleUpdateTableOptions, tableOptions} = this.props
+    const fixFirstColumn = !tableOptions.fixFirstColumn
+    handleUpdateTableOptions({...tableOptions, fixFirstColumn})
+  }
+
+  public handleFieldUpdate = field => {
+    const {handleUpdateTableOptions, tableOptions, dataLabels} = this.props
+    const {sortBy, fieldNames} = tableOptions
+    const fields =
+      fieldNames.length >= dataLabels.length
+        ? fieldNames
+        : this.computedFieldNames
+    const updatedFields = fields.map(
+      f => (f.internalName === field.internalName ? field : f)
+    )
+    const updatedSortBy =
+      sortBy.internalName === field.internalName
+        ? {...sortBy, displayName: field.displayName}
+        : sortBy
+
+    handleUpdateTableOptions({
+      ...tableOptions,
+      fieldNames: updatedFields,
+      sortBy: updatedSortBy,
+    })
+  }
+
+  public componentWillMount() {
+    const {handleUpdateTableOptions, tableOptions} = this.props
+    handleUpdateTableOptions({
+      ...tableOptions,
+      fieldNames: this.computedFieldNames,
+    })
+  }
+
+  public shouldComponentUpdate(nextProps) {
+    const {tableOptions, dataLabels} = this.props
+    const tableOptionsDifferent = !_.isEqual(
+      tableOptions,
+      nextProps.tableOptions
+    )
+    const dataLabelsDifferent = !_.isEqual(dataLabels, nextProps.dataLabels)
+
+    return tableOptionsDifferent || dataLabelsDifferent
+  }
+
+  public render() {
+    const {
+      tableOptions: {timeFormat, fieldNames, verticalTimeAxis, fixFirstColumn},
+      onResetFocus,
+      tableOptions,
+    } = this.props
+
+    const tableSortByOptions = this.computedFieldNames.map(field => ({
+      key: field.internalName,
+      text: field.displayName || field.internalName,
+    }))
+
+    const fields = fieldNames.length > 1 ? fieldNames : this.computedFieldNames
```

Comment:
Switching back and forth between fieldNames and computedFieldNames is confusing. What is implicitly being expressed here?